src/main/java/com/puppycrawl/tools/checkstyle/site/PropertiesMacro.java

```diff
@@ -0,0 +1,493 @@
+            throw new MacroExecutionException(message, checkstyleException);
+        }
+    }
+
+    /**
+     * Writes the properties table for the given module. Expects that the module has been processed
+     * with the ClassAndPropertiesSettersJavadocScraper before calling this method.
+     *
+     * @param sink the sink to write to.
+     * @throws MacroExecutionException if an error occurs during writing.
+     */
+    private static void writePropertiesTable(XdocSink sink)
+            throws MacroExecutionException {
+        sink.table();
+        sink.setInsertNewline(false);
+        sink.tableRows(null, false);
+        sink.rawText(INDENT_LEVEL_12);
+        writeTableHeaderRow(sink);
+        writeTablePropertiesRows(sink);
+        sink.rawText(INDENT_LEVEL_10);
+        sink.tableRows_();
+        sink.table_();
+        sink.setInsertNewline(true);
+    }
+
+    /**
+     * Writes the table header row with 5 columns - name, description, type, default value, since.
+     *
+     * @param sink sink to write to.
+     */
+    private static void writeTableHeaderRow(Sink sink) {
+        sink.tableRow();
+        writeTableHeaderCell(sink, "name");
+        writeTableHeaderCell(sink, "description");
+        writeTableHeaderCell(sink, "type");
+        writeTableHeaderCell(sink, "default value");
+        writeTableHeaderCell(sink, "since");
+        sink.rawText(INDENT_LEVEL_12);
+        sink.tableRow_();
+    }
+
+    /**
+     * Writes a table header cell with the given text.
+     *
+     * @param sink sink to write to.
+     * @param text the text to write.
+     */
+    private static void writeTableHeaderCell(Sink sink, String text) {
+        sink.rawText(INDENT_LEVEL_14);
+        sink.tableHeaderCell();
+        sink.text(text);
+        sink.tableHeaderCell_();
+    }
+
+    /**
+     * Writes the rows of the table with the 5 columns - name, description, type, default value,
+     * since. Each row corresponds to a property of the module.
+     *
+     * @param sink sink to write to.
+     * @throws MacroExecutionException if an error occurs during writing.
+     */
+    private static void writeTablePropertiesRows(Sink sink)
+            throws MacroExecutionException {
+        final Map<String, DetailNode> unmodifiableJavadocs =
+                ClassAndPropertiesSettersJavadocScraper.getJavadocs();
+        final Map<String, DetailNode> javadocs = new LinkedHashMap<>(unmodifiableJavadocs);
+        final Object instance = SiteUtil.getModuleInstance(currentModuleName);
+        final Class<?> clss = instance.getClass();
+
+        final Set<String> properties = SiteUtil.getProperties(clss);
+        final Set<String> allProperties = SiteUtil
+                .fixCapturedProperties(instance, clss, properties);
+
+        for (String property : allProperties) {
+            javadocs.putIfAbsent(property, PARENT_MODULES_PROPERTIES_JAVADOCS.get(property));
```

Comment:
please move this putIfAbsent to block that calculates allProperties.

we need keep code in phases: data preparation, execution of writes.

Reply:
Done.